### PR TITLE
refactor: separate deployer into serval api and implement query and scale

### PR DESCRIPTION
**Pull request by shawnh2**
related to #160

in the following PR, the `pkg/cmd`, `pkg/deployer`, `pkg/connector` shall be removed or rewrote.

### CODE REVIEW

File: cmd/gtctl/cluster_scale.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+	if s.ComponentType != string(greptimedbclusterv1alpha1.FrontendComponentKind) &&
+		s.ComponentType != string(greptimedbclusterv1alpha1.DatanodeComponentKind) &&
+		s.ComponentType != string(greptimedbclusterv1alpha1.MetaComponentKind) {
+		return fmt.Errorf("component type is invalid")
+	}
+
+	if s.Replicas < 1 {
+		return fmt.Errorf("replicas should be equal or greater than 1")
+	}
```
Comment (daviderli614) [inline cmd/gtctl/cluster_scale.go]: maybe replicas can equal 0 ?

Comment (shawnh2) [reply cmd/gtctl/cluster_scale.go]: sure, set replicas to 0, means no need to create any more pod

**Comment by zyy17** [comment]
What a big PR #_#, let me slow down and have a look.

**Comment by zyy17** [comment]
I have another abstraction design as the following structs that can be more clean:

```go
type Cluster struct { ... }

type ClusterOperation interface {
    Create(ctx context.Context, name string, options *CreateOptions) error
    Get(ctx context.Context, name string, options *GetOptions) (*Cluster, error)
    List(ctx context.Context, options *ListOptions) ([]*Cluster, error)
    Update(ctx context.Context, options *UpdateOptions)
    Connect(ctx context.Context, options *ConnectOptions)
}
```

We can implement the K8s and bare-metal clusteroption and provide the factory method. For K8s mode, the creation of an etcd cluster and operator can be hidden in `Create()`.

 
For this PR, IMHO:

1. `pkg/api`: I think it's better to maintain in `pkg/cluster`. `api` is more like the interface for cli users. But in this scenario, it's the apis for internal usage.
2. `pkg/api/scale` and `pkg/api/query`: maybe the two apis are too small, and it's unnecessary to split them from the parent interface. Keep in one interface will be clean;

**Comment by shawnh2** [comment]
> I have another abstraction design as the following structs that can be more clean:
> 
> ```go
> type Cluster struct { ... }
> 
> type ClusterOperation interface {
>     Create(ctx context.Context, name string, options *CreateOptions) error
>     Get(ctx context.Context, name string, options *GetOptions) (*Cluster, error)
>     List(ctx context.Context, options *ListOptions) ([]*Cluster, error)
>     Update(ctx context.Context, options *UpdateOptions)
>     Connect(ctx context.Context, options *ConnectOptions)
> }
> ```
> 
> We can implement the K8s and bare-metal clusteroption and provide the factory method. For K8s mode, the creation of an etcd cluster and operator can be hidden in `Create()`.
> 
> For this PR, IMHO:
> 
> 1. `pkg/api`: I think it's better to maintain in `pkg/cluster`. `api` is more like the interface for cli users. But in this scenario, it's the apis for internal usage.
> 2. `pkg/api/scale` and `pkg/api/query`: maybe the two apis are too small, and it's unnecessary to split them from the parent interface. Keep in one interface will be clean;

great idea, made some changes, PTAL @zyy17

**Review by zyy17 — APPROVED** [review]
LGTM

**Pull request merged**